numpy/core/_internal.py

```diff
@@ -433,6 +433,37 @@ def _copy_fields(ary):
                   'formats': [dt.fields[name][0] for name in dt.names]}
     return array(ary, dtype=copy_dtype, copy=True)
 
+def _promote_fields(dt1, dt2):
+    """ Perform type promotion for two structured dtypes.
+
+    Parameters
+    ----------
+    dt1 : structured dtype
+        First dtype.
+    dt2 : structured dtype
+        Second dtype.
+
+    Returns
+    -------
+    out : dtype
+        The promoted dtype
+
+    Notes
+    -----
+    This function ignores the "titles"
+    """
+    if (dt1.names is None or dt2.names is None) or dt1.names != dt2.names:
+        raise TypeError("invalid type promotion")
+
+    # If the lengths all match, assume the titles (if existing) match the
+    # names.
+    if not len(dt1.names) == len(dt1.fields) == len(dt2.fields):
+        raise TypeError(
+            "NumPy currently does not support promotion with field titles.")
+
+    return dtype([(name, promote_types(dt1[name], dt2[name]))
+                  for name in dt1.names])
```

Comment:
Jeesh, I realized that this is actually incorrect.  What is missing is something like `align=dt1.isalignedstruct or dt2.isalignedstruct`.

Is that what we want? Preserve alignment if one is aligned, or drop alignment when one is not aligned?

Reply:
I agree with your choice of preserving alignment when one is aligned.